### PR TITLE
History real data

### PR DESCRIPTION
**Pull request by Envl**
Retrieved book and user data to display in single history message.
Fixed minor style issues.

### CODE REVIEW

File: src/components/History/Components.js [hunk 1]
```diff
@@ -29,49 +29,84 @@ ItemInfo.propTypes = {
...
+  // supplement: 'Duration: 1-23 Mar. 2019'
 };
 
 const RequestMessage = props => {
   const [id, setId] = useState(props.message.id);
+  const [showOperation, setShowOperation] = useState(
+    props.message.status === 'ongoing'
+  );
+  const [status, setStatus] = useState(props.message.status);
+  const {createdAt} = props.message.stuff
```
Comment (bt-s) [inline src/components/History/Components.js]: Haha I get that we are sort of in a hurry, but could you please rename ```stuff``` to something a little more intelligible?

File: src/components/History/Components.js [hunk 2]
```diff
@@ -29,49 +29,84 @@ ItemInfo.propTypes = {
...
+          : null}
+      </div>
       <div className="msg-body">
         <ItemInfo
-          imgURL={props.message.imgURL}
-          title={props.message.title}
-          status={props.message.status}
+          imgURL={
+            props.message.stuff
+              ? props.message.stuff.imageUrls.length > 0
```
Comment (bt-s) [inline src/components/History/Components.js]: Let me tell you something cool. I think that his can be rewritten to something like the following:

``` props.message.stuff && props.message.stuff.imageUrls.length > 0 && props.message.stuff.imageUrls[0]```

I have to go right now, so I cannot test it, but maybe give it a try.

For reference see https://stackoverflow.com/questions/2932754/ternary-operators-in-javascript-without-an-else

File: src/components/History/Components.js [hunk 3]
```diff
@@ -29,49 +29,84 @@ ItemInfo.propTypes = {
...
-              className="btn accept"
-              onClick={() =>
-                props.acceptCallback ? props.acceptCallback(id) : null
-              }>
-              Accept
-            </button>
-          </div>
+          {showOperation && (
+            <div className="operation">
+              <button
```
Comment (bt-s) [inline src/components/History/Components.js:81]: Could you use the ```Button``` component here?

File: src/pages/HistoryPage.js [hunk 4]
```diff
@@ -28,18 +29,82 @@ const fakeHistories = [
 
 const HistoryPage = props => {
   const [msgType, setMsgType] = useState('give');
-  const histories = fakeHistories; // This should be retrieved on page load
+  const [gotTransactions, setGotTransactions] = useState(false);
+  const [transactions, setTransactions] = useState([]);
+
+  // const histories = fakeHistories; // This should be retrieved on page load
+  console.log('got?', gotTransactions);
```
Comment (bt-s) [inline src/pages/HistoryPage.js]: I think that these two lines can be deleted.

File: src/pages/HistoryPage.js [hunk 5]
```diff
@@ -28,18 +29,82 @@ const fakeHistories = [
...
+
+  if (!gotTransactions) {
+    props.firebase
+      .transactions()
+      .get()
+      .then(querySnapshot => {
+        const transacs = querySnapshot.docs.map(doc => {
+          return {id: doc.id, ...doc.data()};
+        });
+        transacs.forEach(transac => {
```
Comment (bt-s) [inline src/pages/HistoryPage.js:20]: Using ```map()``` is probably faster here.

File: src/pages/HistoryPage.js [hunk 6]
```diff
@@ -28,18 +29,82 @@ const fakeHistories = [
...
+          props.firebase
+            .user(
+              transac.providerID === props.firebase.getMyUID()
+                ? transac.consumerID
+                : transac.providerID
+            )
+            .get()
+            .then(user => {
+              if (user.exists) {
```
Comment (bt-s) [inline src/pages/HistoryPage.js]: Here you could use the following ternary expression:

```let transac.involvedUser = user.exists ? user.data() : false;```

File: src/pages/HistoryPage.js [hunk 7]
```diff
@@ -28,18 +29,82 @@ const fakeHistories = [
...
+                transac.involvedUser = user.data();
+              } else {
+                transac.involvedUser = false;
+              }
+            });
+          props.firebase
+            .book(transac.itemID)
+            .get()
+            .then(book => {
+              if (book.exists) {
```
Comment (bt-s) [inline src/pages/HistoryPage.js]: Same here:

```let transac.stuff = book.exists ? book.data() : false;```

File: src/pages/HistoryPage.js [hunk 8]
```diff
@@ -28,18 +29,82 @@ const fakeHistories = [
...
+                transac.stuff = book.data();
+              } else {
+                transac.stuff = false;
+              }
+            });
+        });
+        console.log('missing data added', transacs);
```
Comment (bt-s) [inline src/pages/HistoryPage.js:36]: This line can be deleted.

File: src/pages/HistoryPage.js [hunk 9]
```diff
@@ -28,18 +29,82 @@ const fakeHistories = [
...
+
+        // 我得在这里 把missing 的info 加入到transactions中, 这些信息需要根据userid 去获取
```
Comment (bt-s) [inline src/pages/HistoryPage.js:38]: Hahahahaha

**Review by bt-s — APPROVED** [review]
This looks way better than it did before, Ming!

I currently do not have the time to test it extensively, but we can do that tomorrow.

**EDIT:**

I just saw in my console that there are some warnings. Maybe you could fix these before merging:

```
./src/components/History/Components.js
  Line 37:  'id' is assigned a value but never used     no-unused-vars
  Line 37:  'setId' is assigned a value but never used  no-unused-vars

./src/pages/HistoryPage.js
  Line 7:  'fakeHistories' is assigned a value but never used  no-unused-vars

./src/pages/BookDetail.js
  Line 7:    'Link' is defined but never used                no-unused-vars
  Line 8:   'ROUTES' is defined but never used              no-unused-vars
  Line 9:    'GoogleApiWrapper' is defined but never used    no-unused-vars
  Line 21:  'loading' is assigned a value but never used    no-unused-vars
  Line 72:   'latitude' is assigned a value but never used   no-unused-vars
  Line 73:   'longitude' is assigned a value but never used  no-unused-vars
```

File: src/pages/HistoryPage.js (hunk 5 above)
Comment (Envl) [reply src/pages/HistoryPage.js:20]: map won't change the original data which is my intent.

File: src/pages/HistoryPage.js (hunk 7 above)
Comment (Envl) [reply src/pages/HistoryPage.js]: Yep, that's more concise.

File: src/pages/HistoryPage.js (hunk 9 above)
Comment (Envl) [reply src/pages/HistoryPage.js:38]: That's a spell.

File: src/components/History/Components.js (hunk 3 above)
Comment (Envl) [reply src/components/History/Components.js:81]: I find it less intuitive to me.

File: src/components/History/Components.js (hunk 1 above)
Comment (Envl) [reply src/components/History/Components.js]: I meant literally stuff. But changed due to "book" is our convention to name a "stuff"

File: src/components/History/Components.js (hunk 2 above)
Comment (Envl) [reply src/components/History/Components.js]: That's way clearer. Changed.

**Pull request merged**